lib/actions/api.js

```diff
@@ -17,32 +19,113 @@ export const nonRealtimeRoutingResponse = createAction('NON_REALTIME_ROUTING_RES
+export const forgetSearch = createAction('FORGET_SEARCH')
+
+function formatRecentPlace (place) {
+  return {
+    ...place,
+    type: 'recent',
+    icon: 'clock-o',
+    id: `recent-${randId()}`,
+    timestamp: new Date().getTime()
+  }
+}
 
-let lastSearchId = 0
+function formatRecentSearch (url, otpState) {
+  return {
+    query: getTripOptionsFromQuery(otpState.currentQuery, true),
+    url,
+    id: randId(),
+    timestamp: new Date().getTime()
+  }
+}
+
+function isStoredPlace (place) {
+  return ['home', 'work', 'suggested', 'stop'].indexOf(place.type) !== -1
+}
+
+/**
+ * Compute the initial activeItinerary. If this is the first search of
+ * session (i.e. searches lookup is empty/null) AND an activeItinerary ID
+ * is specified in URL parameters, use that ID. Otherwise, use null/0.
+ */
+function getActiveItinerary (otpState) {
+  let activeItinerary = otpState.currentQuery.routingType === 'ITINERARY' ? 0 : null
+  // We cannot use window.history.state here to check for the active
+  // itinerary param because it is unreliable in some states (e.g.,
+  // when the print layout component first loads).
+  const urlParams = getUrlParams()
+  if (
+    (!otpState.searches || Object.keys(otpState.searches).length === 0) &&
+    urlParams.ui_activeItinerary
+  ) {
+    activeItinerary = +urlParams.ui_activeItinerary
+  }
+  return activeItinerary
+}
 
-export function routingQuery () {
+/**
+ * Send a routing query to the OTP backend.
+ *
+ * NOTE: We need a random ID so that when a user reloads the page (clearing the
+ * state), performs searches, and presses back to load previous searches
+ * that are no longer contained in the state we don't confuse the search IDs
+ * with search IDs from the new session. If we were to use sequential numbers
+ * as IDs, we would run into this problem.
+ */
+export function routingQuery (searchId = null) {
   return async function (dispatch, getState) {
     const otpState = getState().otp
+    const isNewSearch = !searchId
+    if (isNewSearch) searchId = randId()
     const routingType = otpState.currentQuery.routingType
-    const searchId = ++lastSearchId
-
-    if (!queryIsValid(otpState)) return
-    dispatch(routingRequest({ routingType, searchId }))
+    // Don't permit a routing query if the query is invalid
+    if (!queryIsValid(otpState)) {
+      console.warn('Query is invalid. Aborting routing query', otpState.currentQuery)
+      return
+    }
+    const activeItinerary = getActiveItinerary(otpState)
+    dispatch(routingRequest({ activeItinerary, routingType, searchId }))
 
     // fetch a realtime route
-    fetch(constructRoutingQuery(otpState))
+    const query = constructRoutingQuery(otpState)
+    fetch(query)
       .then(getJsonAndCheckResponse)
       .then(json => {
         dispatch(routingResponse({ response: json, searchId }))
+        // If tracking is enabled, store locations and search after successful
+        // search is completed.
+        // TODO recent searches
```

Comment:
What does this TODO mean?

Reply:
It's old/irrelevant now. It was a reminder to remember recent searches. I'll remove.